File: nocternal-playz/android/ai_assistant/src/main/kotlin/com/nocternal/playz/ai/AssistantEngine.kt
package com.nocternal.playz.ai

import com.nocternal.playz.fx.EnhancerMode
import com.nocternal.playz.model.AudioSource
import com.nocternal.playz.model.Mood
import com.nocternal.playz.model.Playlist
import com.nocternal.playz.model.PlaylistKind
import com.nocternal.playz.theme.EqPresets
import com.nocternal.playz.theme.GenreCatalog
import com.nocternal.playz.theme.GenreDetector
import com.nocternal.playz.theme.ThemePresets

/**
 * The Nocternal AI assistant. Commands are understood and executed on-device ([CommandParser] + the
 * recommenders); open questions go to the optional [AssistantLlm] (Claude) when the user has configured it.
 * The engine never touches the player directly — it returns [AssistantAction]s for the app to run.
 */
class AssistantEngine(
    private val llm: AssistantLlm? = null,
    /** Plugin commands get a chance before the LLM, e.g. "start pomodoro". */
    private val pluginHandler: (String) -> String? = { null },
    private val genres: GenreDetector = GenreDetector(),
    private val moods: MoodDetector = MoodDetector(genres),
    private val recommender: RecommendationEngine = RecommendationEngine(genres, moods),
    private val art: AlbumArtGenerator = AlbumArtGenerator(genres),
) {
    private val conversation = ArrayDeque<LlmTurn>()

    suspend fun handle(text: String, ctx: AssistantContext): AssistantResponse {
        val intent = CommandParser.parse(text)
        if (intent is AssistantIntent.Unknown) {
            pluginHandler(text)?.let { return AssistantResponse(it) }
            return askLlm(text, ctx)
        }
        return execute(intent, ctx)
    }

    fun execute(intent: AssistantIntent, ctx: AssistantContext): AssistantResponse = when (intent) {
        is AssistantIntent.PlayGenre -> playGenre(intent.genreId, ctx)
        is AssistantIntent.PlayMood -> {
            val p = recommender.suggest(ctx, RecommendationEngine.Criteria(mood = intent.mood, hourOfDay = ctx.hourOfDay))
            val genre = genres.forMood(intent.mood, ctx.hourOfDay)
            if (p.trackIds.isEmpty()) searchOnline("${intent.mood.label} music", genre.id)
            else AssistantResponse("Here’s a ${intent.mood.label.lowercase()} mix — ${p.trackIds.size} songs.", listOf(AssistantAction.PlayQueue(p, genre.id)), listOf("Make it longer", "Boost bass", "Sleep in 30 min"))
        }
        is AssistantIntent.PlaySearch -> {
            val (q, src) = intent.query.split('@').let { it[0] to (it.getOrNull(1)?.let(AudioSource::valueOf) ?: AudioSource.LOCAL) }
            val local = ctx.tracks.filter { t -> q.split(' ').all { w -> "${t.title} ${t.artist} ${t.album}".lowercase().contains(w) } }
            if (src == AudioSource.LOCAL && local.isNotEmpty()) {
                AssistantResponse("Playing “${local.first().title}”${if (local.size > 1) " and ${local.size - 1} more" else ""}.",
                    listOf(AssistantAction.PlayQueue(Playlist("ai_search", "Search: $q", local.map { it.id }, PlaylistKind.AI))))
            } else {
                val target = if (src == AudioSource.LOCAL) AudioSource.YOUTUBE else src
                AssistantResponse("Not in your library — searching ${target.label} for “$q”.", listOf(AssistantAction.SwitchSource(target), AssistantAction.Search(q, target)))
            }
        }
        is AssistantIntent.SuggestPlaylist -> {
            val p = recommender.suggest(ctx, RecommendationEngine.Criteria(intent.genreId, intent.mood, intent.bpmMin, intent.bpmMax, ctx.hourOfDay))
            if (p.trackIds.isEmpty()) AssistantResponse("I couldn’t find matching songs in your library yet. Want me to look on YouTube Music or the Radio Hub?", suggestions = listOf("Search YouTube", "Open Radio Hub"))
            else AssistantResponse("I made “${p.name}” with ${p.trackIds.size} songs. ${p.description}.", listOf(AssistantAction.PlayQueue(p, p.genreId)), listOf("Save playlist", "Auto-mix it", "Shuffle"))
        }
        is AssistantIntent.AdjustBass -> AssistantResponse(
            if (intent.up) "Bass boosted. The limiter keeps it clean." else "Bass reduced.",
            listOf(AssistantAction.ChangeBass(if (intent.up) 0.2f else -0.2f)), listOf("More bass", "Reset EQ"),
        )
        is AssistantIntent.ApplyEqPreset -> AssistantResponse("EQ set to ${EqPresets.byId(intent.presetId).name}.", listOf(AssistantAction.SetEq(intent.presetId)))
        AssistantIntent.RecommendEq, AssistantIntent.OptimizeEq -> {
            val advice = EqAdvisor.recommend(ctx.nowPlaying, ctx.route, genres)
            AssistantResponse("${advice.reason}. Applied “${advice.preset.name}”.", listOf(AssistantAction.SetEq(advice.preset.id)), listOf("Boost bass", "Vocal clarity", "Flat"))
        }
        is AssistantIntent.ActivateTheme -> {
            val g = GenreCatalog.byId(intent.genreId)!!
            AssistantResponse("${ThemePresets.byId(g.themePresetId).name} activated ${g.emoji}", listOf(AssistantAction.ApplyGenreTheme(g.id)), listOf("Play ${g.displayName}", "Lighting settings"))
        }
        is AssistantIntent.SwitchSource -> AssistantResponse("Switched to ${intent.source.label}.", listOf(AssistantAction.SwitchSource(intent.source)))
        is AssistantIntent.SleepTimer -> AssistantResponse("Sleep timer set for ${formatMinutes(intent.minutes)}. I’ll fade the music out gently.", listOf(AssistantAction.SetSleepTimer(intent.minutes)), listOf("Play sleep sounds", "Cancel timer"))
        is AssistantIntent.Transport -> AssistantResponse(transportReply(intent.command), listOf(AssistantAction.Transport(intent.command)))
        is AssistantIntent.Enhance -> AssistantResponse("${intent.mode.label} on: ${intent.mode.description.replaceFirstChar { it.lowercase() }}.", listOf(AssistantAction.Enhance(intent.mode)), EnhancerMode.entries.filter { it != intent.mode }.take(2).map { it.label })
        is AssistantIntent.VocalRemover -> AssistantResponse(if (intent.on) "Karaoke mode on — vocals removed. Lyrics are on the player screen." else "Vocals are back.", listOf(AssistantAction.SetVocalRemover(if (intent.on) 1f else 0f)))
        is AssistantIntent.ExplainFeature -> AssistantResponse(FeatureExplainer.explain(intent.topic))
        AssistantIntent.IdentifySong -> AssistantResponse("Listening… hold your phone near the music.", listOf(AssistantAction.StartRecognition))
        AssistantIntent.GenerateAlbumArt -> ctx.nowPlaying?.let { AssistantResponse("Here’s neon art for “${it.title}”.", listOf(AssistantAction.ShowAlbumArt(art.generate(it)))) }
            ?: AssistantResponse("Play a song first and I’ll design its cover.")
        AssistantIntent.AutoMix -> AssistantResponse("DJ auto-mix on: I’ll pick key- and tempo-matched songs and blend them on the beat.", listOf(AssistantAction.EnableAutoMix(true)), listOf("Turn off auto-mix"))
        is AssistantIntent.Unknown -> AssistantResponse(helpText(ctx))
    }

    private fun playGenre(genreId: String, ctx: AssistantContext): AssistantResponse {
        val g = GenreCatalog.byId(genreId) ?: return AssistantResponse(helpText(ctx))
        val p = recommender.suggest(ctx, RecommendationEngine.Criteria(genreId = genreId, hourOfDay = ctx.hourOfDay))
        if (p.trackIds.isEmpty()) return searchOnline(g.aiSuggestions.first(), g.id)
        return AssistantResponse(
            "Playing your ${g.displayName} playlist ${g.emoji} — ${ThemePresets.byId(g.themePresetId).name} theme on.",
            listOf(AssistantAction.PlayQueue(p, g.id), AssistantAction.ApplyGenreTheme(g.id)),
            g.aiSuggestions.drop(1).take(2),
        )
    }

    private fun searchOnline(query: String, genreId: String): AssistantResponse = AssistantResponse(
        "Nothing like that in your library yet, so I’m tuning the Radio Hub to “$query”.",
        listOf(AssistantAction.ApplyGenreTheme(genreId), AssistantAction.SwitchSource(AudioSource.RADIO), AssistantAction.Search(query, AudioSource.RADIO)),
        listOf("Try YouTube Music instead"),
    )

    private suspend fun askLlm(text: String, ctx: AssistantContext): AssistantResponse {
        val client = llm ?: return AssistantResponse(helpText(ctx))
        val system = buildSystemPrompt(ctx)
        conversation.addLast(LlmTurn(LlmRole.USER, text))
        while (conversation.size > 12) conversation.removeFirst()
        return try {
            val reply = client.complete(system, conversation.toList())
            conversation.addLast(LlmTurn(LlmRole.ASSISTANT, reply))
            // The model can ask us to run a command by ending with a line "ACTION: <command>".
            val actionLine = reply.lines().lastOrNull { it.startsWith("ACTION:") }
            val clean = reply.lines().filterNot { it.startsWith("ACTION:") }.joinToString("\n").trim()
            val actions = actionLine?.removePrefix("ACTION:")?.trim()?.let { cmd ->
                CommandParser.parse(cmd).takeIf { it !is AssistantIntent.Unknown }?.let { execute(it, ctx).actions }
            }.orEmpty()
            AssistantResponse(clean, actions, fromLlm = true)
        } catch (e: Exception) {
            conversation.removeLast()
            AssistantResponse("I couldn’t reach the AI service (${e.message ?: "network error"}). Commands like “play lo-fi” still work offline.")
        }
    }

    private fun buildSystemPrompt(ctx: AssistantContext): String = buildString {
        appendLine("You are the assistant inside NOCTERNAL PLAYZ, a neon music player app. Answer in 1–4 short sentences, friendly and specific.")
        appendLine("You can control the app by ending your reply with one line \"ACTION: <command>\" using one of: play <genre|mood|song>, boost bass, reduce bass, recommend eq, activate <genre> theme, switch to <local|youtube|radio>, sleep in <n> minutes, enhance clarity, remove noise, karaoke on, auto mix, identify song. Only add an ACTION if the user wants something done.")
        appendLine("Built-in genres: ${GenreCatalog.all.joinToString { it.displayName }}.")
        appendLine("Time of day: ${ctx.hourOfDay}:00. Source: ${ctx.source.label}. Output: ${ctx.route.name.lowercase()}.")
        ctx.nowPlaying?.let { t ->
            append("Now playing: “${t.title}” by ${t.artist}")
            t.bpm?.let { append(", ${it.toInt()} BPM") }
            t.camelotKey?.let { append(", key $it") }
            genres.detect(t)?.let { append(", genre ${it.genre.displayName}") }
            appendLine(".")
        }
        if (!ctx.privateMode && ctx.tracks.isNotEmpty()) {
            val topGenres = ctx.tracks.mapNotNull { genres.detect(it)?.genre?.displayName }.groupingBy { it }.eachCount().entries.sortedByDescending { it.value }.take(5)
            appendLine("Library: ${ctx.tracks.size} songs; top genres ${topGenres.joinToString { "${it.key} (${it.value})" }}.")
        }
    }

    private fun helpText(ctx: AssistantContext) = buildString {
        append("Try “play trance playlist”, “boost bass”, “activate meditation theme”, “sleep in 30 minutes” or “what song is this?”.")
        if (llm == null) append(" Add a Claude API key in Settings → AI to ask me anything else.")
    }

    private fun transportReply(c: TransportCommand) = when (c) {
        TransportCommand.PLAY -> "Playing."
        TransportCommand.PAUSE -> "Paused."
        TransportCommand.NEXT -> "Next song."
        TransportCommand.PREVIOUS -> "Going back."
        TransportCommand.SHUFFLE -> "Shuffle on."
        TransportCommand.REPEAT -> "Repeat on."
        TransportCommand.VOLUME_UP -> "Louder."
        TransportCommand.VOLUME_DOWN -> "Quieter."
        TransportCommand.LIKE -> "Added to Favorites 💜"
    }

    private fun formatMinutes(m: Int) = if (m >= 60 && m % 60 == 0) "${m / 60} h" else if (m >= 60) "${m / 60} h ${m % 60} min" else "$m min"

    companion object {
        /** Mood → a short line the Home screen shows when the AI detects the listening mood. */
        fun moodLine(mood: Mood) = "Feeling ${mood.label.lowercase()}? I matched the lights."
    }
}

File: nocternal-playz/ios/NocternalKit/Sources/AIAssistant/AssistantEngine.swift
import Foundation
import NocternalModel
import ThemeManager
import FXEngine

public struct LLMTurn: Sendable { public enum Role: String, Sendable { case user, assistant }; public let role: Role; public let text: String }

public protocol AssistantLLM: Sendable {
    func complete(system: String, turns: [LLMTurn]) async throws -> String
}

/// Claude backend over raw HTTPS (there is no official Swift SDK). The key is the user's own and stays in
/// the Keychain; for a public release, proxy requests through your own server instead.
public struct ClaudeLLM: AssistantLLM {
    let apiKey: String
    let model: String
    public init(apiKey: String, model: String = "claude-opus-5") { self.apiKey = apiKey; self.model = model }

    public func complete(system: String, turns: [LLMTurn]) async throws -> String {
        var req = URLRequest(url: URL(string: "https://api.anthropic.com/v1/messages")!)
        req.httpMethod = "POST"
        req.timeoutInterval = 60
        req.setValue(apiKey, forHTTPHeaderField: "x-api-key")
        req.setValue("2023-06-01", forHTTPHeaderField: "anthropic-version")
        req.setValue("server-side-fallback-2026-07-01", forHTTPHeaderField: "anthropic-beta")
        req.setValue("application/json", forHTTPHeaderField: "content-type")
        let body: [String: Any] = [
            "model": model,
            "max_tokens": 4096,
            "system": system,
            "output_config": ["effort": "low"],
            "fallbacks": "default",
            "messages": turns.map { ["role": $0.role.rawValue, "content": $0.text] },
        ]
        req.httpBody = try JSONSerialization.data(withJSONObject: body)
        let (data, resp) = try await URLSession.shared.data(for: req)
        guard let obj = try JSONSerialization.jsonObject(with: data) as? [String: Any] else { throw URLError(.cannotParseResponse) }
        if let http = resp as? HTTPURLResponse, http.statusCode != 200 {
            let msg = (obj["error"] as? [String: Any])?["message"] as? String ?? "HTTP \(http.statusCode)"
            throw NSError(domain: "Claude", code: http.statusCode, userInfo: [NSLocalizedDescriptionKey: msg])
        }
        if obj["stop_reason"] as? String == "refusal" { return "I can’t help with that one — ask me about music, playlists or sound settings." }
        let blocks = obj["content"] as? [[String: Any]] ?? []
        let text = blocks.compactMap { $0["type"] as? String == "text" ? $0["text"] as? String : nil }.joined(separator: "\n")
        return text.isEmpty ? "Sorry, I didn’t catch that." : text
    }
}

/// The Nocternal AI assistant: offline commands first, optional Claude for everything else.
public final class AssistantEngine {
    let llm: AssistantLLM?
    let pluginHandler: (String) -> String?
    let recommender = RecommendationEngine()
    private var conversation: [LLMTurn] = []

    public init(llm: AssistantLLM?, pluginHandler: @escaping (String) -> String? = { _ in nil }) { self.llm = llm; self.pluginHandler = pluginHandler }

    public func handle(_ text: String, _ ctx: AssistantContext) async -> AssistantResponse {
        let intent = CommandParser.parse(text)
        if case .unknown = intent {
            if let r = pluginHandler(text) { return AssistantResponse(r) }
            return await askLLM(text, ctx)
        }
        return execute(intent, ctx)
    }

    public func execute(_ intent: AssistantIntent, _ ctx: AssistantContext) -> AssistantResponse {
        switch intent {
        case .playGenre(let id):
            guard let g = GenreCatalog.byId(id) else { return AssistantResponse(help) }
            let p = recommender.suggest(ctx, genreId: id, mood: nil, bpmMin: nil, bpmMax: nil)
            if p.trackIds.isEmpty { return radioFallback(g.aiSuggestions.first ?? g.displayName, g.id) }
            return AssistantResponse("Playing your \(g.displayName) playlist \(g.emoji) — \(ThemePresets.byId(g.themePresetId).name) theme on.",
                                     [.playQueue(p, genreId: g.id), .applyGenreTheme(g.id)], suggestions: Array(g.aiSuggestions.dropFirst().prefix(2)))
        case .playMood(let m):
            let p = recommender.suggest(ctx, genreId: nil, mood: m, bpmMin: nil, bpmMax: nil)
            let g = GenreDetector().forMood(m, hour: ctx.hour)
            return p.trackIds.isEmpty ? radioFallback("\(m.label) music", g.id)
                : AssistantResponse("Here’s a \(m.label.lowercased()) mix — \(p.trackIds.count) songs.", [.playQueue(p, genreId: g.id)], suggestions: ["Boost bass", "Sleep in 30 min"])
        case .playSearch(let q, let src):
            let words = q.split(separator: " ").map(String.init)
            let local = ctx.tracks.filter { t in words.allSatisfy { "\(t.title) \(t.artist) \(t.album)".lowercased().contains($0) } }
            if src == .local, let first = local.first {
                return AssistantResponse("Playing “\(first.title)”.", [.playQueue(Playlist(id: "ai_search", name: "Search: \(q)", trackIds: local.map(\.id), kind: .ai), genreId: nil)])
            }
            let target: AudioSource = src == .local ? .youtube : src
            return AssistantResponse("Searching \(target.label) for “\(q)”.", [.switchSource(target), .search(q, target)])
        case let .suggestPlaylist(genreId, mood, lo, hi):
            let p = recommender.suggest(ctx, genreId: genreId, mood: mood, bpmMin: lo, bpmMax: hi)
            return p.trackIds.isEmpty ? AssistantResponse("No matches in your library yet — want me to search YouTube Music or the Radio Hub?", suggestions: ["Open Radio Hub"])
                : AssistantResponse("I made “\(p.name)” with \(p.trackIds.count) songs.", [.playQueue(p, genreId: p.genreId)], suggestions: ["Auto mix", "Shuffle"])
        case .adjustBass(let up): return AssistantResponse(up ? "Bass boosted. The limiter keeps it clean." : "Bass reduced.", [.changeBass(up ? 0.2 : -0.2)])
        case .recommendEq:
            let (p, why) = EqAdvisor.recommend(ctx.nowPlaying, route: ctx.route)
            return AssistantResponse("\(why). Applied “\(p.name)”.", [.setEq(p)])
        case .activateTheme(let id):
            let g = GenreCatalog.byId(id)!
            return AssistantResponse("\(ThemePresets.byId(g.themePresetId).name) activated \(g.emoji)", [.applyGenreTheme(id)])
        case .switchSource(let s): return AssistantResponse("Switched to \(s.label).", [.switchSource(s)])
        case .sleepTimer(let m): return AssistantResponse("Sleep timer set for \(m) min. I’ll fade the music out gently.", [.setSleepTimer(m)])
        case .transport(let c): return AssistantResponse(c == .like ? "Added to Favorites 💜" : "Done.", [.transport(c)])
        case .enhance(let m): return AssistantResponse("\(m.label) on.", [.enhance(m)])
        case .vocalRemover(let on): return AssistantResponse(on ? "Karaoke mode on — vocals removed." : "Vocals are back.", [.setVocalRemover(on ? 1 : 0)])
        case .explain(let topic): return AssistantResponse(FeatureExplainer.explain(topic))
        case .identifySong: return AssistantResponse("Listening… hold your phone near the music.", [.startRecognition])
        case .generateArt:
            guard let t = ctx.nowPlaying else { return AssistantResponse("Play a song first and I’ll design its cover.") }
            return AssistantResponse("Here’s neon art for “\(t.title)”.", [.showAlbumArt(.generate(for: t))])
        case .autoMix: return AssistantResponse("DJ auto-mix on: key- and tempo-matched songs, blended on the beat.", [.enableAutoMix(true)])
        case .unknown: return AssistantResponse(help)
        }
    }

    private var help: String {
        "Try “play trance playlist”, “boost bass”, “activate meditation theme”, “sleep in 30 minutes” or “what song is this?”." + (llm == nil ? " Add a Claude API key in Settings to ask me anything else." : "")
    }

    private func radioFallback(_ q: String, _ genreId: String) -> AssistantResponse {
        AssistantResponse("Nothing like that in your library yet, so I’m tuning the Radio Hub to “\(q)”.", [.applyGenreTheme(genreId), .switchSource(.radio), .search(q, .radio)])
    }

    private func askLLM(_ text: String, _ ctx: AssistantContext) async -> AssistantResponse {
        guard let llm else { return AssistantResponse(help) }
        conversation.append(LLMTurn(role: .user, text: text))
        if conversation.count > 12 { conversation.removeFirst(conversation.count - 12) }
        var system = "You are the assistant inside NOCTERNAL PLAYZ, a neon music player. Answer in 1–4 short sentences. You can control the app by ending with one line \"ACTION: <command>\" (play <genre|mood|song>, boost bass, recommend eq, activate <genre> theme, switch to <local|youtube|radio>, sleep in <n> minutes, karaoke on, auto mix, identify song). Only add an ACTION if the user wants something done. Built-in genres: \(GenreCatalog.all.map(\.displayName).joined(separator: ", ")). Time: \(ctx.hour):00."
        if let t = ctx.nowPlaying { system += " Now playing: “\(t.title)” by \(t.artist)." }
        do {
            let reply = try await llm.complete(system: system, turns: conversation)
            conversation.append(LLMTurn(role: .assistant, text: reply))
            let lines = reply.components(separatedBy: "\n")
            let action = lines.last { $0.hasPrefix("ACTION:") }.map { String($0.dropFirst(7)).trimmingCharacters(in: .whitespaces) }
            let clean = lines.filter { !$0.hasPrefix("ACTION:") }.joined(separator: "\n").trimmingCharacters(in: .whitespacesAndNewlines)
            var actions: [AssistantAction] = []
            if let a = action { let intent = CommandParser.parse(a); if case .unknown = intent {} else { actions = execute(intent, ctx).actions } }
            return AssistantResponse(clean, actions, fromLLM: true)
        } catch {
            conversation.removeLast()
            return AssistantResponse("I couldn’t reach the AI service (\(error.localizedDescription)). Commands like “play lo-fi” still work offline.")
        }
    }
}
